contracts/ColonyNetworkAuction.sol

```diff
@@ -0,0 +1,203 @@
+    require(endBlock > 0);
+    _;
+  }
+
+  modifier allBidsClaimed  {
+    require(tokensClaimed == receivedTotal);
+    _;
+  }
+
+  event AuctionStarted(uint indexed _startBlock);
+  event AuctionBid(address indexed _sender, uint _amount, uint _missingFunds);
+  event AuctionClaim(address indexed _recipient, uint _sentAmount);
+  event AuctionFinalized(uint _finalPrice);
+
+  function DutchAuction(
+    address _clnyToken,
+    address _token,
+    uint256 _quantity,
+    uint _startPrice) 
+    public 
+    {
+    colonyNetwork = msg.sender;
+    require(_clnyToken != 0x0 && _token != 0x0);
+    clnyToken = ERC20Extended(_clnyToken);
+    token = ERC20Extended(_token);
+    
+    require(_startPrice > 0);
+    startPrice = _startPrice;
+    require(_quantity > 0);
+    quantity = _quantity;
+    startBlock = block.number;
+    AuctionStarted(startBlock);
+  }
+
+  function bid(uint256 _amount) public
+  auctionNotFinalized
+  endBlockNotSet
+  {
+    require(_amount > 0);
+    uint _totalToEndAuction = totalToEndAuction();
+    uint remainingToEndAuction = sub(_totalToEndAuction, receivedTotal);
+
+    // Adjust the amount for final bid in case that takes us over the offered quantity at current price
+    // Also conditionally set the endBlock
+    uint amount;
+    if (remainingToEndAuction > _amount) {
+      amount = _amount;
+    } else {
+      amount = remainingToEndAuction;
+      endBlock = block.number;
+    }
+    
+    clnyToken.transferFrom(msg.sender, this, amount);
+    bids[msg.sender] = add(bids[msg.sender], amount);
+    receivedTotal = add(receivedTotal, amount);
+    
+    AuctionBid(msg.sender, amount, sub(remainingToEndAuction, amount));
+  }
+
+  // Finalize the auction and sets the final Token price
+  function finalize() public
+  auctionNotFinalized
+  endBlockSet
+  {
+    // Give the network all CLNY sent to the auction in bids
+    clnyToken.transfer(colonyNetwork, receivedTotal);
+    finalPrice = mul(receivedTotal, MULTIPLIER) / quantity;
+    finalized = true;
+    AuctionFinalized(finalPrice);
+  }
+
+  function claim() public 
+  auctionFinalized
+  returns (bool)
+  {
+    uint amount = bids[msg.sender];
+    uint tokens = mul(amount, MULTIPLIER) / finalPrice;
+
+    // Due to finalPrice floor rounding, the number of assigned tokens may be higher
+    // than expected. Therefore, the number of remaining unassigned auction tokens
+    // may be smaller than the number of tokens needed for the last claimTokens call
+    uint auctionTokenBalance = token.balanceOf(address(this));
+    if (tokens > auctionTokenBalance) {
+      tokens = auctionTokenBalance;
+    }
+
+    // Update the total amount of funds for which tokens have been claimed
+    tokensClaimed += amount;
+
+    // Set receiver bid to 0 before transferring the tokens
+    bids[msg.sender] = 0;
+    require(token.transfer(msg.sender, tokens));
+    AuctionClaim(msg.sender, tokens);
+
+    assert(token.balanceOf(msg.sender) >= tokens);
```

Comment:
Given we're talking about BYOT, what do you think about enforcing that these token balances are correct?